File: BAEKJOON/graph_search/1012.py
"""
문제 제목 : 유기농 배추
문제 난이도 : 하
문제 유형 : DFS, BFS

연결 요소의 개수를 세는 문제
모든 정점에 대하여 DFS 및 BFS를 수해앟고, 한 번 방문한 정점은 다시 확인하지 않습니다.
전체적으로 DFS 및 BFS를 수행한 총 횟수를 계산
DFS 및 BFS 응용 문제 중에서 출제 비중이 매우 높은 유형 중 하나
DFS로 문제를 푸는 경우, sys 라이브러리의 setrecursionlimit() 함수 설정을 해줄 필요가 있음
"""

import sys
sys.setrecursionlimit(100000)
def dfs(x,y):
    visited[x][y]=True
    directions = [(-1,0),(1,0),(0,-1),(0,1)]
    for dx, dy in directions:
        nx, ny = x+dx, y+dy
        if nx<0 or ny>=m or ny<0 or nx>=n:
            continue
        if array[nx][ny] and not visited[nx][ny]:
            dfs(nx,ny)

for _ in range(int(input())):
    m,n,k=map(int,input().split())
    array = [[0]*m for _ in range(n)]
    visited = [[False]*m for _ in range(n)]
    for _ in range(k):
        y, x = map(int, input().split())
        array[x][y]=1
    result=0
    for i in range(n):
        for j in range(m):
            if array[i][j] and not visited[i][j]:
                dfs(i,j)
                result+=1
    print(result)

File: BAEKJOON/graph_search/dfs/1012.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().split("\n");
let t = Number(input[0]);
let line = 1;

while (t--) {
  const [m, n, k] = input[line].split(" ").map(Number);
  const graph = [];
  let result = 0;
  for (let i = 0; i < n; i++) {
    graph[i] = new Array(m);
  }
  for (let i = 1; i <= k; i++) {
    let [y, x] = input[line + i].split(" ").map(Number);
    graph[x][y] = 1;
  }
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (dfs(graph, n, m, i, j)) {
        result += 1;
      }
    }
  }
  line += k + 1;
  console.log(result);
}

function dfs(graph, n, m, x, y) {
  if (x < 0 || y < 0 || x >= n || y >= m) {
    return false;
  }
  if (graph[x][y] === 1) {
    graph[x][y] = -1;
    dfs(graph, n, m, x - 1, y);
    dfs(graph, n, m, x, y - 1);
    dfs(graph, n, m, x + 1, y);
    dfs(graph, n, m, x, y + 1);
    return true;
  }
  return false;
}
